src/host/VtIo.cpp

```diff
@@ -373,148 +267,392 @@ void VtIo::SetWindowVisibility(bool showOrHide) noexcept
-[[nodiscard]] HRESULT VtIo::ManuallyClearScrollback() const noexcept
+// SGR 1006: SGR Extended Mouse Mode
+void VtIo::WriteSGR1006(bool enabled)
+{
+    char buf[] = "\x1b[?1003;1006h";
+    buf[std::size(buf) - 2] = enabled ? 'h' : 'l';
+    WriteUTF8({ buf, std::size(buf) - 1 });
+}
+
+// DECAWM: Autowrap Mode
+void VtIo::WriteDECAWM(bool enabled)
+{
+    char buf[] = "\x1b[?7h";
+    buf[std::size(buf) - 2] = enabled ? 'h' : 'l';
+    WriteUTF8({ buf, std::size(buf) - 1 });
+}
+
+// LNM: Line Feed / New Line Mode
+void VtIo::WriteLNM(bool enabled)
+{
+    char buf[] = "\x1b[20h";
+    buf[std::size(buf) - 2] = enabled ? 'h' : 'l';
+    WriteUTF8({ buf, std::size(buf) - 1 });
+}
+
+// ASB: Alternate Screen Buffer
+void VtIo::WriteASB(bool enabled)
 {
-    if (_pVtRenderEngine)
+    char buf[] = "\x1b[?1049h";
+    buf[std::size(buf) - 2] = enabled ? 'h' : 'l';
+    WriteUTF8({ buf, std::size(buf) - 1 });
+}
+
+void VtIo::WriteAttributes(WORD attributes)
+{
+    FormatAttributes(_back, attributes);
+    _flush();
+}
+
+void VtIo::WriteInfos(til::point target, std::span<const CHAR_INFO> infos)
+{
+    const auto beg = infos.begin();
+    const auto end = infos.end();
+    const auto last = end - 1;
+    const auto cork = Cork();
+    WORD attributes = 0xffff;
+
+    WriteCUP(target);
+
+    for (auto it = beg; it != end; ++it)
     {
-        return _pVtRenderEngine->ManuallyClearScrollback();
+        const auto& ci = *it;
+        auto ch = ci.Char.UnicodeChar;
+        auto wide = WI_IsAnyFlagSet(ci.Attributes, COMMON_LVB_LEADING_BYTE | COMMON_LVB_TRAILING_BYTE);
+
+        if (wide)
+        {
+            if (WI_IsAnyFlagSet(ci.Attributes, COMMON_LVB_LEADING_BYTE))
+            {
+                if (it == last)
+                {
+                    // The leading half of a wide glyph won't fit into the last remaining column.
+                    // --> Replace it with a space.
+                    ch = L' ';
+                    wide = false;
+                }
+            }
+            else
+            {
+                if (it == beg)
+                {
+                    // The trailing half of a wide glyph won't fit into the first column. It's incomplete.
+                    // --> Replace it with a space.
+                    ch = L' ';
+                    wide = false;
+                }
+                else
+                {
+                    // Trailing halves of glyphs are ignored within the run. We only emit the leading half.
+                    continue;
+                }
+            }
+        }
+
+        if (attributes != ci.Attributes)
+        {
+            attributes = ci.Attributes;
+            WriteAttributes(attributes);
+        }
+
+        const auto isSurrogate = til::is_surrogate(ch);
+        const auto isControl = IsControlCharacter(ch);
+        int repeat = 1;
+        if (isSurrogate || isControl)
+        {
+            ch = isSurrogate ? UNICODE_REPLACEMENT : L' ';

```

Comment:
I think this is might break legacy apps that deliberately write C0 controls to the buffer to get glyphs from the DOS 437 code page. We've only recently fixed that in WT. Can we not retain that mapping here? (see PR #16825)

Reply:
I was just about to finish coding this and noticed that the 437 codepage defined even more glyphs: https://en.wikipedia.org/wiki/Code_page_437#Character_set
Shouldn't we translate those too instead of converting them to "?"?

Reply:
Oh, and we could of course also translate them according to the current code page instead of always using CP437. Is that something we'd want?

Reply:
Eh, you know what... Ignore me. Let's go with what we had. :)

Reply:
If you mean the C1 range, the reason I didn't do them was because they didn't seem to be supported in the v1 console either (they were just rendering as some form of replacement glyph). I don't think it would necessarily be a bad thing to add them, but I'm not sure it's worthwhile if there aren't any existing applications using them.

The other thing to bear in mind is that the C1 range was locale specific, whereas the C0 glyphs were standard across all DOS code pages (at least that was my understanding).

Reply:
Didn't see you follow up messages. Ignore me. :)